### PR TITLE
Prevent loss of access to data with auto-configuration for legacy Redis databases

### PR DESCRIPTION
**Pull request by delano**
## Summary

Fixes loss of access to data issue affecting some (likely most) Docker users upgrading to v0.23. When existing installations used minimal environment variables, they relied on hardcoded defaults that distributed data across Redis databases 1, 6, 7, 8, and 11. With v0.23's shift to defaulting everything to database 0, these installations would lose access to their existing data without any indication of the problem.

The solution introduces auto-configuration that detects legacy data distribution and automatically configures the application to preserve access while providing a clear path toward database consolidation.

## Problem Context

**Affected users**: Docker installations using only `SECRET` and `REDIS_URI` environment variables
**Risk**: Complete loss of access to sessions, customers, secrets, and other data
**Timing**: During v0.23 upgrade when new defaults take effect

Without this fix, the application would silently connect to database 0 while user data remained inaccessible in legacy databases, creating a hidden migration scenario with no warning.

### Auto-Configuration Logic
- **Detection**: Scans Redis databases 1-15 for existing data patterns
- **Analysis**: Determines if current configuration uses defaults that would cause data loss  
- **Response**: Automatically updates `OT.conf['redis']['dbs']` to preserve access to detected legacy databases
- **Communication**: Clear messaging about what's happening and next steps

### User Experience Changes

**Case A: Explicit Configuration** (unchanged)
```
LEGACY DATA DETECTED - No action required
Found existing data in legacy databases:
  • 500 session records in database 1 [was legacy default]
✅ Continuing with existing configuration
```

**Case B: Minimal Environment Variables** (now protected)
```
LEGACY DATA DETECTED - Auto-configuring for compatibility  
Found existing data in legacy databases:
  • 500 session records in database 1
Auto-configured database mappings to preserve data access
To migrate to database 0 (recommended before v1.0): `bin/ots migrate-redis-data --run`
```

## Security & Compatibility

- **Backward compatible**: No impact on existing explicit configurations
- **Forward compatible**: Clear path to v1.0 database consolidation  
- **Non-destructive**: Never modifies or moves data automatically
- **Transparent**: Logs all auto-configuration decisions for auditability

## Testing

- **15 test cases** covering detection logic, auto-configuration scenarios, and edge cases
- **Demo script** (`scripts/test_auto_config_demo.rb`) for ongoing validation
- **Migration tool compatibility** maintained with new detection format

## Migration Path

Users get three clear options without pressure:
1. **No action** - current setup continues working with auto-configuration
2. **Migrate later** - use provided tooling when convenient  
3. **See documentation** - comprehensive guide for permanent configuration

This ensures seamless v0.23 upgrades while establishing the foundation for eventual v1.0 database consolidation.

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: Claude <noreply@anthropic.com>

___

### **PR Type**
Enhancement

___

### **Description**
- Add Redis legacy data detection and auto-configuration system

- Implement migration tool for database consolidation

- Prevent silent data loss during v0.23+ upgrades

- Update documentation and Docker references

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Startup Detection"] --> B["Scan Redis DBs 1-15"]
  B --> C["Legacy Data Found?"]
  C -->|Yes| D["Auto-configure DB mappings"]
  C -->|No| E["Continue with defaults"]
  D --> F["Preserve data access"]
  G["Migration Tool"] --> H["Preview changes"]
  H --> I["Execute migration"]
  I --> J["Consolidate to DB 0"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>detect_legacy_data.rb</strong><dd><code>Add comprehensive legacy data detection system</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-39d5885d87a4c7925aa18d6ba98a029e7fe7517c2762e86cd89dcbbc46b9c4a9">+240/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>migrate_redis_data_command.rb</strong><dd><code>Implement Redis data migration CLI tool</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-15fe85e72c67dda76cb8a86f7a87a24a8166971098d01c8838b6c859951ce5e3">+150/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>connect_databases.rb</strong><dd><code>Integrate legacy data detection into startup</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-e99578753339bad16ce8309531f1d7e71940af4c95a1af3fae9263f844f011e8">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>boot.rb</strong><dd><code>Skip log banner for CLI commands</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-65e08ff06f3b888e6ee7b0c7df20e80d0ffb98792a6f6db4d9b6001e32516284">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>cli.rb</strong><dd><code>Register new migrate-redis-data command</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-682e856ca4b6e4f9b379fa54b0f348a75da65b5f589f9da4b23de0f49cd6d1bb">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ots</strong><dd><code>Add migrate-redis-data command and improve output</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-f760b4e89ed099696292fc69306124cbec19b6c57ac9be1f5ff22bc4b887171b">+8/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>detect_legacy_data_try.rb</strong><dd><code>Add comprehensive test coverage for detection</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-40993b7871d58e8577cdf134e036efecb840ec141779586800aa9f7aa4a65a11">+260/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>test_auto_config_demo.rb</strong><dd><code>Add demo script for auto-configuration behavior</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-096c7fcc204f3e4bea43cf3ba6c23014d28003c1e8fa8b84219fe677ce622d22">+85/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>12 files</summary><table>
<tr>
  <td><strong>redis-migration.md</strong><dd><code>Add comprehensive Redis migration guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-6c98727cffb2fd047db954dba0cb9b416d6ebd017d62d90e9c2805339d803214">+197/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>development.md</strong><dd><code>Add development guide with debugging tips</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-97db29a7915320e63d41d38a0440360a87055ee8ed03757aa263116dbbb4aabe">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>docker.md</strong><dd><code>Add comprehensive Docker deployment guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-ae5611658dcd443d7a30575bebcdf15a71b7a0f5abebb18b62d00aa0373cae74">+193/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>docker-lite.md</strong><dd><code>Add Docker lite image documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-67d226694f9227c4e7f68b12199d683c67de075f7c45fdbd52fc75ac90455164">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>postman.md</strong><dd><code>Add Postman API development guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-63314b26116651749ae694dd00ed9495e065e9605a3adfc20ef3fd6c0c1bcca1">+209/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Update with v0.23 upgrade information</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+16/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Dockerfile</strong><dd><code>Update documentation reference paths</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Dockerfile-lite</strong><dd><code>Update documentation reference paths</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-243eac714cadfdb5ff11f1c3e2f29c0066af9f08b0dcdda44bfcac262968eed0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>DEVELOPMENT.md</strong><dd><code>Update installation guide reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-d9657425e036fb0ad2912c2c24ea793648eb86de5b40aa0844276dd08ed01d87">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>DOCKER-lite.md</strong><dd><code>Update cross-reference to docker.md</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-440cfd661738e00a6593e41a39517d9839282d065208f379cf43b1453e81d5f0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>conf.py</strong><dd><code>Add Sphinx documentation configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-85933aa74a2d66c3e4dcdf7a9ad8397f5a7971080d34ef1108296a7c6b69e7e3">+29/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Update documentation reference paths</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-291f8b8d9125baac658c8fa7b6f4df9066abd996262166e5a6fa05eb4de0a351">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>LICENSE.txt</strong><dd><code>Update copyright notice format</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-d0ed4cc3fb70489fe51c7e0ac180cba2a7472124f9f9e9ae67b01a37fbd580b7">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>.keep</strong></td>
  <td><a href="https://github.com/onetimesecret/onetimesecret/pull/1731/files#diff-c63b836a7589a083d20c25ab171761c9a73363dba2c11487a2eb3d55f8fcfbdf">[link]</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: lib/onetime/cli/migrate_redis_data_command.rb [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+      puts "\n🧹 Cleaning up migration flags..."
+      legacy_data.each do |model, locations|
+        locations.each do |location|
+            client   = Familia.dbclient(location[:database])
+            flag_key = Familia.join(['ots', 'migration_needed', model, "db_#{location[:database]}"])
+            client.del(flag_key)
+            puts "   Removed migration flag: #{flag_key}"
+        rescue StandardError => ex
+            OT.ld "Could not remove migration flag for #{model}: #{ex.message}"
+        end
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/onetime/cli/migrate_redis_data_command.rb:157]: It's a loop rescue and not a begin/rescue block. It'll catch exceptions for everything in the loop.

File: lib/onetime/initializers/detect_legacy_data.rb [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+      # installations, auto-configuring to legacy databases preserves data access.
+      # However, for new installations, this would perpetuate the old multi-database
+      # configuration we're trying to migrate away from. Since we cannot distinguish
+      # between these scenarios, the safest approach is to require manual
+      # intervention via explicit configuration or environment flags.
+      raise ex
+
+    # Determines if the current database configuration is using all defaults (database 0)
+    # which means auto-configuration is needed to preserve access to legacy data
+    def using_default_database_config?(current_dbs, legacy_locations)
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/onetime/initializers/detect_legacy_data.rb:91]: This was caused by a syntax error right above. See [cd1a390](https://github.com/onetimesecret/onetimesecret/pull/1731/commits/cd1a390696f07ac9ddb76493260e4101ead3a185).
